[27828] Feature: Query menu in sidenav

ingamai opened the pull request:
Query menu (originally as dropdown on click of the title) should be moved to the left sidenav.

### Done:
- [x] Insert query dropdown component into the sidebar (delete dependencies to the OpContextMenu for this is not longer a dropdown context menu)
- [x] Menu should be opened all the time and show all queries (Autocompleter is searching for an empty string on init, which shows all available queries) 
- [x] Load menu on init once and make it accessible from every point in the project (listen if submenu work-packages has been opened)
- [x] Define default queries and insert them in the autocompleter to make them searchable 
- [x] Set title of default queries manually (They cannot be found, because they have not query, so currently they are recognized by their number and type of filters, which is not ideal but working)
- [x] Insert menu item "Summary" into default queries (-> special handling needed because it has it's own url ".../report" -> currently reload needed
- [x] Sort queries by new categories in right order
- [x] Make categories foldable and style new menu
- [x] Highlight current item in menu
- [x] Synchronize menu and view (menu should listen on changes, e.g. renaming, deletion, creation of queries, and update immeadiatly)
- [x] Adapt wording of query settings context menu to changes (Publish -> Visibility settings)
- [x] Make query title inline editable and change title in menu immediatly -> delete settings menu item "Rename"
- [x] Adapt wording of settings menu
- [x] Inline edit: 
   - [x] Check if field empty (throw error and color input field red)
   - [x] Check if new name of query already exists
   - [x] Disable input fields of default queries

### To-Do's

- [x] Query "All open" is opened on initialisation, but not highlighted
- [x] Correct highlighting after title changing error
- [x] Change icon of "Visibility settings"

- [x] Harmonize font sizes and margines with menu items in Wiki
- [x] Static queries (e.g. "Latest activity", "Created by me", "Assigned to me") do not change the state (i.e. url) and by that the user is not able to reload the page.  
- [x] Highlight default queries correctly after reload 
- [x] All global menus without hamburger icon should be open on default / after reload
- [ ] ( Add Label of page at top of the menu; missing in pages: )
   - Work Packages
   - Administration
   - My account
- [x] ( Make menu accessible by keyboard )
- [x] Add indicator when query has changed

### Plugins

- [x] Reporting: https://github.com/finnlabs/openproject-reporting/pull/128

### Work packages
https://community.openproject.com/projects/openproject/work_packages/27828/activity

wielinde committed:
in main menu add gantt as extra work package child item

wielinde committed:
Satisfy spec and code climate

wielinde committed:
Add gantt chart icon behin default gantt query name.

Commit:
Merge branch 'feature/add-gantt-to-menus' into feature/query-menu-in-sidenav

Commit:
WIP Query menu in left sidebar

Commit:
Merge branch 'angular-cli' of https://github.com/opf/openproject into feature/query-menu-in-sidenav

Commit:
Shift query dropdown in left sidenav

Commit:
Reload menu or load query on click from every project location

Commit:
WIP set correct label for default queries

Commit:
Query menu listens on all changes of queries (delete, create, rename, toggle starred) and updates immediatly

ingamai committed:
Merge dev in local branch 'feature/query-menu-in-sidenav'

ingamai committed:
WIP: Inline edit, field validation

ingamai committed:
Inline Edit validation and comfirm

ingamai committed:
Merge branch 'dev' into feature/query-menu-in-sidenav

ingamai committed:
Inline edit: validation of duplicate name

ingamai committed:
Set default columns and sorting for static queries

ingamai committed:
Codeclimate issues fixed

ingamai committed:
WIP Inline edit validation not working perfectly in all error states

ingamai committed:
Inline edit working

ingamai committed:
Autocompleter hover disabled and hovering over categories fixed

ingamai committed:
Category hover and toggle fixed; tested in Chrome, Firefox and Opera

ingamai committed:
Placeholder cut off fixed and text wrap added

ingamai committed:
English and german wording adjusted

ingamai committed:
Styles of inline edit and menu adjusted; matching wiki page styles

ulferts committed:
Merge branch 'dev' into feature/query-menu-in-sidenav

ulferts committed:
prevent menus to be displayed to often

* application menu only displayed on work package

ulferts committed:
specify using no_menu layout more often

ulferts committed:
adapt tests to altered production implementation

ingamai committed:
Hamburger icon only in project; on global wp page: default queries shown correctly and summary removed

ingamai committed:
Merge branch 'feature/query-menu-in-sidenav' of https://github.com/opf/openproject into feature/query-menu-in-sidenav

ulferts committed:
searching for undefined leads to error

ingamai committed:
Merge branch 'feature/query-menu-in-sidenav' of https://github.com/opf/openproject into feature/query-menu-in-sidenav

ulferts committed:
load menu on wp summary page

ingamai committed:
Accessible click fixed (listen on escape)

ingamai committed:
Gantt in top menu deleted (gantt chart is part of default queries on wp page)

ulferts committed:
reduce times queries are loaded

ulferts committed:
lowercase on second word

ulferts committed:
remove menu from search and home

ingamai committed:
Styles fixed (category toggle and correct highlighting)

ulferts committed:
reflect static query in url

ulferts committed:
fix autocomplete handling in specs

ingamai committed:
Open all global menus on default and hide hamburger icon on global pages; Rebuild changes that have been ovrwritten after merge"

ingamai committed:
Correct highlighting of default queries  after reload

oliverguenther committed:
Replace summary cuke with spec

oliverguenther commented:
Here's my list of things I've found:

- [x] Meta/Alt/Ctrl clicking a menu item no longer works. You cant force an item to open in a new tab.
- [x] Clicking on a menu item, changing a filter, then clicking on the item again does not properly update the `query_props` , however does correctly show the query. This causes some tests to break.
- [x] The duplicate name check is not enforced in the backend. It's perfectly fine to save two queries with the same name with 'Save as'. So I'd suggest we remove the duplicate check.
- [x] All items are being refreshed twice
  - To load all entries in the duplicate check.
  - To load all entries when re-rendering. Also: This might take quite some time when lots of queries are visible.
- [x] When overriding the query_props (e.g., (remove a filter), the selectable-title becomes empty but clickable 
![selectable-title](https://user-images.githubusercontent.com/459462/43703820-404570d2-995e-11e8-925e-a301e75fdc10.gif)

- [x] I believe we can remove the `wp-query-menu.service` completely. It previously controlled reloading the menu on updating / saving / renaming / removing a query.
- [x] Searching for a query does not highlight / select the first match
- [x] Clicking in the middle of the editable title does not set position
- [x] Query props are no longer correctly updated

oliverguenther committed:
WIP

oliverguenther committed:
Clear up selectors

oliverguenther committed:
Avoid actively setting promises and instead use $state.go to load links

[ci skip]

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into feature/query-menu-in-sidenav

[ci skip]

oliverguenther committed:
Make editable title component a little simpler

We can reuse the component I built for the wiki, that wasn't present in
the frontend beforehand.

oliverguenther committed:
Fix moving through the menu and selecting items

[ci skip]

oliverguenther committed:
Add save button to query title when query changed

oliverguenther committed:
Improve static names lookup by comparing query_props

oliverguenther committed:
Adapt and fix specs

oliverguenther committed:
Allow inner scrolling of wp query results

Also, style the webkit scrollbar to make it pretty where supported

oliverguenther committed:
Allow renaming the query through setting menu, but simply focus on field
